File: python/problem-12.py
import math

def tri_nums():
    i, num = 1, 1
    while True:
        yield num
        i += 1
        num += i

def divisors(num):
    divisors = []
    for i in range(1, int(math.sqrt(num) + 1)):
        if num % i is 0:
            divisors.append(i)
            if i is not num/i:
                divisors.append(int(num/i))
    return divisors

print(next(filter(lambda x: len(divisors(x)) > 500, tri_nums())))

File: clojure/problem_12.clj
(def tri-nums (reductions + (range)))

(defn sqrt [n]
  (first (filter #(> (* % %) n) (range))))

(defn factors [n]
  (mapcat
    #(if-not (= % (/ n %)) [% (/ n %)] [%])
    (filter #(zero? (mod n %)) (range 1 (sqrt n)))))

(println (first (filter
                  #(> (count (factors %)) 500)
                  tri-nums))) ;; => 76576500
